torchaudio/models/_wavernn.py

```diff
@@ -85,3 +87,106 @@ def forward(self, specgram: Tensor) -> Tensor:
+
+
+class _UpsampleNetwork(nn.Module):
+    r"""Upscale the dimensions of a spectrogram to match waveform.
+
+    Args:
+        upsample_scales: the list of upsample scales
+        n_res_block: the number of ResBlock in stack (default=10)
+        n_freq: the number of bins in a spectrogram (default=128)
+        n_hidden: the number of hidden dimensions (default=128)
+        n_output: the number of output dimensions (default=128)
+        kernel_size: the number of kernel size in the first Conv1d layer (default=5)
+
+    Examples
+        >>> upsamplenetwork = _UpsampleNetwork(upsample_scales=[4, 4, 16])
+        >>> input = torch.rand(10, 128, 10)  # a random spectrogram
+        >>> output = upsamplenetwork(input)  # shape: (10, 1536, 128), (10, 1536, 128)
+    """
+
+    def __init__(self,
+                 upsample_scales: List[int],
+                 n_res_block: int = 10,
+                 n_freq: int = 128,
+                 n_hidden: int = 128,
+                 n_output: int = 128,
+                 kernel_size: int = 5) -> None:
+        super().__init__()
+
+        total_scale = 1
+        for upsample_scale in upsample_scales:
+            total_scale *= upsample_scale
+
+        self.indent = (kernel_size - 1) // 2 * total_scale
+        self.resnet = _MelResNet(n_res_block, n_freq, n_hidden, n_output, kernel_size)
+        self.resnet_stretch = _Stretch2d(total_scale, 1)
+
+        up_layers = []
+        for scale in upsample_scales:
+            stretch = _Stretch2d(scale, 1)
+            conv = nn.Conv2d(in_channels=1,
+                             out_channels=1,
+                             kernel_size=(1, scale * 2 + 1),
+                             padding=(0, scale),
+                             bias=False)
+            conv.weight.data.fill_(1. / (scale * 2 + 1))
+            up_layers.append(stretch)
+            up_layers.append(conv)
+        self.upsample_layers = nn.Sequential(*up_layers)
+
+    def forward(self, specgram: Tensor) -> Tensor:
+        r"""Pass the input through the _UpsampleNetwork layer.
+        Args:
+            specgram (Tensor): the input sequence to the _UpsampleNetwork layer (n_batch, n_freq, n_time)
+
+        Return:
+            Tensor shape: (n_batch, (n_time - kernel_size + 1) * total_scale, n_freq),
+                          (n_batch, (n_time - kernel_size + 1) * total_scale, n_output)
+        where total_scale is the product of all elements in upsample_scales.
+        """
+
+        resnet_output = self.resnet(specgram).unsqueeze(1)
+        resnet_output = self.resnet_stretch(resnet_output)
+        resnet_output = resnet_output.squeeze(1)
+
+        specgram = specgram.unsqueeze(1)
+        upsampling_output = self.upsample_layers(specgram)
+        upsampling_output = upsampling_output.squeeze(1)[:, :, self.indent:-self.indent]
+
+        return upsampling_output.transpose(1, 2), resnet_output.transpose(1, 2)
```

Comment:
nit: I'm not sure whether we should change the convention established in other libraries, but I would not transpose here.

Reply:
How about adding the transpose in two lines before?
```
resnet_output = resnet_output.squeeze(1).transpose()
upsampling_output = upsampling_output.squeeze(1)[:, :, self.indent:-self.indent].transpose()

return upsampling_output, resnet_output
```
Otherwise I has to transpose it in the WaveRNN code because the second dimension has to be same: 
line 288 in #735 as:
```
mels, aux = self.upsample(specgram)
mels = mels.transpose()
aux = aux.transpose()
```

Reply:
As we talked offline, let's move the transpose to WaveRNN instead.

Reply:
I removed the transpose operation here and it will be added after the [line ](https://github.com/pytorch/audio/pull/735/files#diff-4680098e69e34986c5cca1f1c26b4303R288)of WaveRNN PR